pkg/controller/controller_utils.go

```diff
@@ -631,6 +672,86 @@ func maxContainerRestarts(pod *api.Pod) int {
 	return maxRestarts
 }
 
+// ControlActivePods first filters out inactive pods, then it tries to take
+// control of the remaining pods that matching the selector. It returns the pods
+// that are active and are controlled by the controller, and the pods that
+// should be abandoned by the controller.
+func ControlActivePods(podControl PodControlInterface,
+	controllerObject api.ObjectMeta,
+	controllerSelector labels.Selector,
+	controllerKind unversioned.GroupVersionKind,
+	pods []api.Pod,
+) (controlledPods []*api.Pod, toAbandonPods []*api.Pod) {
+	for i := range pods {
+		pod := pods[i]
+		if !IsPodActive(pod) {
+			glog.V(4).Infof("Ignoring inactive pod %v/%v in state %v, deletion time %v",
+				pod.Namespace, pod.Name, pod.Status.Phase, pod.DeletionTimestamp)
+			continue
+		}
+		controlled, controllerRef := ControllerOf(pod.ObjectMeta)
+		if controlled {
+			if controllerRef.UID == controllerObject.UID {
+				// already controlled
+				if controllerSelector.Matches(labels.Set(pod.Labels)) {
+					controlledPods = append(controlledPods, &pod)
+				} else {
+					toAbandonPods = append(toAbandonPods, &pod)
+				}
+			} else {
+				// ignoring the pod controlled by other controller
+				glog.V(4).Infof("Ignoring pod %v/%v, it's owned by [%s/%s, name: %s, uid: %s]",
+					pod.Namespace, pod.Name, controllerRef.APIVersion, controllerRef.Kind, controllerRef.Name, controllerRef.UID)
+				continue
+			}
+		} else {
+			if !controllerSelector.Matches(labels.Set(pod.Labels)) {
+				continue
+			}
+			// try to take control
+			if ControlPod(podControl, controllerObject, controllerKind, pod) {
+				// successfully take control
+				controlledPods = append(controlledPods, &pod)
+			} else {
+				// failed to take control, ignore the pod
+				continue
+			}
+		}
+	}
+	return controlledPods, toAbandonPods
+}
+
+// ControllerOf returns if the controllee is controlled by a controller and if
+// so, a reference to the controller.
+func ControllerOf(controllee api.ObjectMeta) (controlled bool, controllerRef api.OwnerReference) {
+	for _, owner := range controllee.OwnerReferences {
+		// controlled by other controller
+		if owner.Controller != nil && *owner.Controller == true {
+			return true, owner
+		}
+	}
+	return false, api.OwnerReference{}
+}
+
+// ControlPod sends a patch to take control of the pod. It returns if it succeeds
+// in taking control.
+func ControlPod(podControl PodControlInterface, controllerObject api.ObjectMeta, controllerKind unversioned.GroupVersionKind, pod api.Pod) bool {
+	// we should not adopt any pods if the controller is about to be deleted
+	if controllerObject.DeletionTimestamp != nil {
```

Comment:
I think don't worry about the race between deletion and adoption, we are more concerned with the race between adoption and orphaning. This `ControlPod` function is called before the rc manager updates rc.Status.ObservedGeneration, so as long as the orphan finalizer waits for the bump of ObservedGeneration, there will be no race between the adoption and orphaning.